lib/gui/app/components/file-selector/file-selector/file-selector.jsx

```diff
@@ -0,0 +1,591 @@
+
+const _ = require('lodash')
+const os = require('os')
+const path = require('path')
+const React = require('react')
+const propTypes = require('prop-types')
+const { default: styled } = require('styled-components')
+const rendition = require('rendition')
+const prettyBytes = require('pretty-bytes')
+const fontAwesome = require('@fortawesome/fontawesome')
+const {
+  faFileAlt,
+  faFolder,
+  faAngleLeft,
+  faHdd
+} = require('@fortawesome/fontawesome-free-solid')
+const Storage = require('../../../models/storage')
+const analytics = require('../../../modules/analytics')
+const middleEllipsis = require('../../../utils/middle-ellipsis')
+const files = require('../../../../../shared/files')
+const selectionState = require('../../../../../shared/models/selection-state')
+
+/**
+ * @summary Recent files localStorage object key
+ * @constant
+ * @private
+ */
+const RECENT_FILES_KEY = 'file-selector-recent-files'
+const recentStorage = new Storage(RECENT_FILES_KEY)
+
+/**
+ * @summary How many directories to show with the breadcrumbs
+ * @type {Number}
+ * @constant
+ * @private
+ */
+const MAX_DIR_CRUMBS = 3
+
+/*
+ * TODO
+ * - move state to redux store, e.g. highlighted file
+ * - make back button
+ * - filters with no hidden files
+ * - 
+ */
+
+/**
+ * @summary Character limit of a filename before a middle-ellipsis is added
+ * @constant
+ * @private
+ */
+const FILENAME_CHAR_LIMIT = 20
+
+/**
+ * @summary Character limit of a filename before a middle-ellipsis is added
+ * @constant
+ * @private
+ */
+const FILENAME_CHAR_LIMIT_SHORT = 15
+
+/**
+ * @summary Color scheme
+ * @constant
+ * @private
+ */
+const colors = {
+  primary: {
+    color: '#3a3c41',
+    background: '#ffffff'
+  },
+  secondary: {
+    color: '#1c1d1e',
+    background: '#ebeff4',
+    title: '#b3b6b9'
+  }
+}
+
+const icons = {
+  faFileAlt: fontAwesome.icon(faFileAlt).html[0],
+  faFolder: fontAwesome.icon(faFolder).html[0],
+  faAngleLeft: fontAwesome.icon(faAngleLeft).html[0],
+  faHdd: fontAwesome.icon(faHdd).html[0]
+}
+
+/**
+ * @summary Get all subpaths contained in a path
+ * @function
+ * @private
+ *
+ * @param {String} fullpath - path string
+ * @returns {Array<Object>} - all subpaths as file objects
+ *
+ * @example
+ * const subpaths = subpaths('/home/user/Downloads')
+ * console.log(subpaths.map(file => file.fullpath))
+ * > [ '/', '/home', '/home/user', '/home/user/Downloads' ]
+ */
+const subpaths = (fullpath) => {
```

Comment:
Move this to `lib/shared/files`. Also pass labels to the functions so we can label generic folders and not just root.